### ISSUE TITLE
Custom theme-based TypeScript Intellisense

### ISSUE THREAD

**Issue by nandorojo** [nandorojo/dripsy#6]
Another goal of this library is to have a script that detects your theme, and builds TypeScript intellisense into your components based on your custom theme values. But this isn't made yet. I'd love help with it.

For example, if you have a theme like this:

```js
export default {
  colors: {
    primary: 'blue',
    muted: '#e8e8e8',
  },
};
```

Then you would get intellisense like this:

<img src="https://media.giphy.com/media/cLMi3vFyXkSgnJal3l/giphy.gif" />

_^ The gif might take sec to load._

I think this could be achieved with some TypeScript wizardry, where a command line takes in the theme file and generates a typescript hidden file in your repository. This hidden file would globally declare values that override the `sx` prop, injecting in type suggestions. The command would probably be placed in a given project's `postinstall`, so that it runs every time you run `yarn` or `npm install`

@slorber recommended I look [here](https://material-ui.com/customization/palette/#adding-new-colors).

I think generating something like this would work?

![Xiy9U6kv](https://user-images.githubusercontent.com/13172299/87368437-a930fb00-c54b-11ea-8f79-04d13d070710.jpg)

**Comment by nandorojo** [comment]
This requires some TypeScript codegen wizardry that I just don't have. I'll probably close this until some TS genius stumbles upon it and can solve it.

**Comment by nandorojo** [comment]
There's an exciting theme UI PR which seems like it could solve this https://github.com/system-ui/theme-ui/pull/1090. Still have to look more into it, but I've re-opened this issue.

**Comment by nandorojo** [comment]
Relevant: https://material-ui.com/components/about-the-lab/#typescript

**Comment by nandorojo** [comment]
After many months, I'm excited to say that this feature is close to landing. @hasparus has been hard at work with many great additions to `theme-ui`, and one of them will address this issue. This means our `sx` prop will get strict types based on our theme.

To follow: https://github.com/system-ui/theme-ui/pull/1090

**Comment by nandorojo** [comment]
#72

**Comment by nandorojo** [comment]
Gosh, this is sooo hard with all of theme-ui's complex types in the way rn.

@cmaycumber I think this is the start of v3 of Dripsy. Remove `theme-ui` altogether, copy over the types ourselves, and have great type safety.

**Comment by nandorojo** [comment]
This cool new project by @intergalacticspacehighway has good custom types: https://github.com/intergalacticspacehighway/react-native-styled-variants/blob/9d6a801683b5ffbc8641b5937b62f24425e504de/src/types.ts#L7

**Comment by cmaycumber** [comment]
@nandorojo Have you seen `react-native-styled-variants`: https://styled-variants.vercel.app/

**Comment by nandorojo** [comment]
Haha yeah good timing. I'd want to make sure that the type creator there from the custom theme is performant. For massive TS objects, you can get into slow performance if you recurse too much (see https://github.com/millsp/ts-toolbelt/issues/154)

**Comment by nandorojo** [comment]
Looks like TS performance has improved for this, which is good news https://github.com/microsoft/TypeScript/pull/45711

**Comment by cmaycumber** [comment]
> Haha yeah good timing. I'd want to make sure that the type creator there from the custom theme is performant. For massive TS objects, you can get into slow performance if you recurse too much (see [millsp/ts-toolbelt#154](https://github.com/millsp/ts-toolbelt/issues/154))

That makes a lot of sense. We should definitely make sure that's the case. I wonder if we can build off of some of the work from `react-native-styled-variants`.

**Comment by nandorojo** [comment]
Yeah, I wrote more about my vision for how this could work at #72 a while ago

**Comment by cmaycumber** [comment]
I think that's definitely a good UX imo. Let me reread #72. I'm super pumped to get this started

**Comment by nandorojo** [comment]
Starting at #124

**Comment by nandorojo** [comment]
<img width="1156" alt="Screen Shot 2021-09-27 at 5 12 04 PM" src="https://user-images.githubusercontent.com/13172299/134986025-84b08662-98b3-4c72-992b-7750d9dab12c.png">

Omg. still work to do here but...

**Comment by nandorojo** [comment]
Adding variant support too...

<img width="702" alt="Screen Shot 2021-09-27 at 6 17 55 PM" src="https://user-images.githubusercontent.com/13172299/134992877-34c5437c-4547-4da4-b3d3-2f84fd0225a8.png">

Theme here (just to mock)

<img width="407" alt="Screen Shot 2021-09-27 at 6 18 13 PM" src="https://user-images.githubusercontent.com/13172299/134992920-a5df7a7e-6ab4-4857-bc60-ee2199505919.png">

**Comment by nandorojo** [comment]
<img width="485" alt="Screen Shot 2021-09-27 at 6 47 19 PM" src="https://user-images.githubusercontent.com/13172299/134995527-d126d611-f224-4267-a665-39ae8457b27a.png">
<img width="386" alt="Screen Shot 2021-09-27 at 6 47 11 PM" src="https://user-images.githubusercontent.com/13172299/134995529-a927645a-c31d-4f30-9710-033bf9a0ca1f.png">

Gradient support working...

**Comment by nandorojo** [comment]
<img width="695" alt="Screen Shot 2021-09-27 at 7 10 32 PM" src="https://user-images.githubusercontent.com/13172299/134997298-2165b8e0-642f-4b4c-b47f-87a66a24eaf9.png">

Added `sx` prop as a function support (not sure why tbh, but why not)

**Comment by nandorojo** [comment]
I published the first version of this at `3.0.0`, would love help testing:

```sh
yarn add dripsy@canary
```

**Comment by nandorojo** [comment]
2 things left:

1. Make sure that the `Tokenizer` is actually performant
2. Maybe add per-scale styles? As in, `bg` only sees scales from the `theme.colors`. It's just a lot of time / effort.

**Comment by nandorojo** [comment]
<img width="388" alt="Screen Shot 2021-09-28 at 6 01 33 PM" src="https://user-images.githubusercontent.com/13172299/135171758-02683278-9ea7-4681-97e4-09eb38772763.png">

Shorthand support working...

**Comment by nandorojo** [comment]
<img width="415" alt="Screen Shot 2021-09-28 at 6 08 16 PM" src="https://user-images.githubusercontent.com/13172299/135172426-d238e343-aaf5-442a-b638-5c20191eef14.png">

Props now narrow down to the specific `themeKey` for a given style 😎

**Comment by rafaelrinaldi** [comment]
Popping in to say I thoroughly appreciate all the work being put into this. It will be such a great addition to Dripsy!

**Comment by nandorojo** [comment]
Thanks, glad you think so! You can see more details at https://github.com/nandorojo/dripsy/pull/124
